activemount.go

```diff
@@ -0,0 +1,152 @@
+		doMountFs = false
+	} else if errors.Is(readDirErr, io.EOF) {
+		// The directory is empty, mount the filesystem
+		doMountFs = true
+	} else {
+		return false, fmt.Errorf("failed to list activemounts/ %v", readDirErr)
+	}
+
+	var activeMountInfo activeMount
+	activemountFilePath := d.activemountsdir(requestName) + requestId
+	file, err := os.Open(activemountFilePath)
+
+	if err == nil {
+		// The file can exist from a previous mount when doing a docker cp on an already mounted container, no need to mount the filesystem again
+		payload, readErr := io.ReadAll(file)
+		closeErr := file.Close()
+		if readErr != nil {
+			return false, fmt.Errorf("active mount file %s has been opened but cannot be read %w", activemountFilePath, readErr)
+		} else if closeErr != nil {
+			return false, fmt.Errorf("active mount file %s has been opened but cannot be closed %w", activemountFilePath, readErr)
+		}
+		unmarshalErr := json.Unmarshal(payload, &activeMountInfo)
+		if unmarshalErr != nil {
+			return false, fmt.Errorf("active mount file %s contents are invalid %w", activemountFilePath, unmarshalErr)
+		}
+	} else if os.IsNotExist(err) {
+		// Default case, we need to create a new active mount, the filesystem needs to be mounted
+		activeMountInfo = activeMount{UsageCount: 0}
+	} else {
+		return false, fmt.Errorf("active mount file %s exists but cannot be opened %w", activemountFilePath, err)
+	}
+
+	activeMountInfo.UsageCount++
+
+	// Convert activeMountInfo to JSON to store it in a file. We can safely ignore Marshal errors, since the
+	// activeMount structure is simple enought not to contain "strage" floats, unsupported datatypes or cycles
+	// which are the error causes for json.Marshal
+	payload, _ := json.Marshal(activeMountInfo)
+	err = os.WriteFile(activemountFilePath, payload, 0o644)
+	if err != nil {
+		return false, fmt.Errorf("active mount file %s cannot be written %w", activemountFilePath, err)
+	}
+
+	return doMountFs, nil
+}
+
+// deactivateVolume checks if the volume that has been requested to be unmounted (as in docker volume unmounting)
+// actually requires to be unmounted (as an overlay fs unmount). It will check the number of times the container
+// has been requested to mount the volume in the file named `requestId` and decrease the number, when the number
+// reaches zero it will delete the `requestId` file since this container no longer mounts the volume. It will
+// also check if other containers are mounting this volume by checking for other files in the active mounts folder.
+// Parameters:
+//
+//	requestName: Name of the volume to be unmounted
+//	requestID: ID of the container requesting the unmount
+//	activemountsdir: Folder where Docker-On-Top mounts are tracked.
+//
+// Return:
+//
+//	doUnmountFs: Returns true if there are not other usages of this volume and the filesystem can be unmounted.
+//	err: If the function encountered an error, the error itself, nil if everything went right.
```

Comment:
```suggestion
//	doUnmountFs: `true` if there are no other usages of this volume and the filesystem shall be unmounted
//		by the caller. **Must be respected even when an error is returned**!
//	err: An error, if encountered, `nil` otherwise.
```

Reply:
I have adapted your message because I have changed the behavior and now it never returns true when there is an error.